flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/Knn.java

```diff
@@ -0,0 +1,211 @@
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * An Estimator which implements the KNN algorithm.
+ *
+ * <p>See: https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm.
+ */
+public class Knn implements Estimator<Knn, KnnModel>, KnnParams<Knn> {
+
+    protected Map<Param<?>, Object> params = new HashMap<>();
+
+    public Knn() {
+        ParamUtils.initializeMapWithDefaultValues(params, this);
+    }
+
+    @Override
+    public KnnModel fit(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        /* Tuple2 : <sampleVector, label> */
+        DataStream<Tuple2<DenseVector, Double>> inputData =
+                tEnv.toDataStream(inputs[0])
+                        .map(
+                                new MapFunction<Row, Tuple2<DenseVector, Double>>() {
+                                    @Override
+                                    public Tuple2<DenseVector, Double> map(Row value) {
+                                        Double label = (Double) value.getField(getLabelCol());
+                                        DenseVector feature =
+                                                (DenseVector) value.getField(getFeaturesCol());
+                                        return Tuple2.of(feature, label);
+                                    }
+                                });
+        DataStream<KnnModelData> distributedModelData = prepareModelData(inputData);
+        DataStream<KnnModelData> modelData = mergeModelData(distributedModelData);
+        KnnModel model = new KnnModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, getParamMap());
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return this.params;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static Knn load(StreamExecutionEnvironment env, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Prepares distributed knn model data. Constructs the sample matrix and computes norm of
+     * features.
+     *
+     * @param inputData Input feature data with label.
+     * @return Distributed knn model.
+     */
+    private static DataStream<KnnModelData> prepareModelData(
```

Comment:
It seems to me that the following implementation is more clear:

We rename `prepareModelData()` to `computeL2Norm()` and let it a simple `MapFunction`.
In this function, we compute the L2 norm of each input data point and output `Tuple3<DenseVector, Double, Double>` (i.e., feature, label, l2norm) for each data point.

Then we rename `mergeModelData` to `genModelData`. We can compact the `features, labels, norms` to `KnnModelData`in the `mapPartitionFunction` function.

In this case, the semantic of `KnnModelData` is unique and only all of the input data points can form the final `model data`. Also we can avoid allocating memory for `KnnModelData` twice.

What do you think?

Reply:
OK，I agree with you.